File: test/language/functions/java/functions/structure_param/StructureParamTest.java
package functions.structure_param;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

import java.io.IOException;
import java.util.Arrays;

import org.junit.Test;

import zserio.runtime.io.ByteArrayBitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamWriter;

public class StructureParamTest
{
    @Test
    public void checkMetresConverterCaller() throws IOException
    {
        final MetresConverterCaller metresConverterCaller = createMetresConverterCaller();
        assertEquals(CONVERTED_CM_VALUE, metresConverterCaller.getCm());

        final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        metresConverterCaller.write(writer);
        final byte[] writtenByteArray = writer.toByteArray();
        writer.close();

        final byte[] expectedByteArray = writeMetresConverterCallerToByteArray();
        assertTrue(Arrays.equals(expectedByteArray, writtenByteArray));

        final ByteArrayBitStreamReader reader = new ByteArrayBitStreamReader(writtenByteArray);
        final MetresConverterCaller readMetresConverterCaller = new MetresConverterCaller(reader);
        assertEquals(metresConverterCaller, readMetresConverterCaller);
    }

    private byte[] writeMetresConverterCallerToByteArray() throws IOException
    {
        ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        writer.writeBits(VALUE_A, 16);
        writer.writeBits(CONVERTED_CM_VALUE, 16);
        writer.close();

        return writer.toByteArray();
    }

    private MetresConverterCaller createMetresConverterCaller()
    {
        final MetresConverterCaller metresConverterCaller = new MetresConverterCaller();

        final MetresConverter metresConverter = new MetresConverter(M_VALUE_TO_CONVERT);
        metresConverter.setA(VALUE_A);

        metresConverterCaller.setMetresConverter(metresConverter);
        metresConverterCaller.setCm(CONVERTED_CM_VALUE);

        return metresConverterCaller;
    }

    private static int VALUE_A = 0xABCD;
    private static int M_VALUE_TO_CONVERT = 2;
    private static int CONVERTED_CM_VALUE = M_VALUE_TO_CONVERT * 100;
}

File: test/language/functions/cpp98/StructureParamTest.cpp
#include <vector>

#include "gtest/gtest.h"

#include "zserio/BitStreamWriter.h"
#include "zserio/BitStreamReader.h"
#include "zserio/ObjectArray.h"

#include "functions/structure_param/MetresConverter.h"
#include "functions/structure_param/MetresConverterCaller.h"

namespace functions
{
namespace structure_param
{

class FunctionsStructureParamTest : public ::testing::Test
{
protected:
    void writeMetresConverterCallerToByteArray(zserio::BitStreamWriter& writer)
    {
        writer.writeBits(VALUE_A, 16);
        writer.writeBits(CONVERTED_CM_VALUE, 16);
    }

    void createMetresConverterCaller(MetresConverterCaller& metresConverterCaller)
    {
        MetresConverter& metresConverter = metresConverterCaller.getMetresConverter();
        metresConverter.setA(VALUE_A);
        metresConverterCaller.setCm(CONVERTED_CM_VALUE);
   }

    static const uint16_t VALUE_A = 0xABCD;
    static const uint16_t M_VALUE_TO_CONVERT = 2;
    static const uint16_t CONVERTED_CM_VALUE = M_VALUE_TO_CONVERT * 100;
};

TEST_F(FunctionsStructureParamTest, checkMetresConverterCaller)
{
    MetresConverterCaller metresConverterCaller;
    createMetresConverterCaller(metresConverterCaller);
    const uint16_t expectedCm = CONVERTED_CM_VALUE;
    ASSERT_EQ(expectedCm, metresConverterCaller.getCm());

    zserio::BitStreamWriter writtenWriter;
    metresConverterCaller.write(writtenWriter);
    size_t writtenWriterBufferByteSize;
    const uint8_t* writtenWriterBuffer = writtenWriter.getWriteBuffer(writtenWriterBufferByteSize);

    zserio::BitStreamWriter expectedWriter;
    writeMetresConverterCallerToByteArray(expectedWriter);
    size_t expectedWriterBufferByteSize;
    const uint8_t* expectedWriterBuffer = expectedWriter.getWriteBuffer(expectedWriterBufferByteSize);

    std::vector<uint8_t> writtenWriterVector(writtenWriterBuffer,
                                             writtenWriterBuffer + writtenWriterBufferByteSize);
    std::vector<uint8_t> expectedWriterVector(expectedWriterBuffer,
                                              expectedWriterBuffer + expectedWriterBufferByteSize);
    ASSERT_EQ(expectedWriterVector, writtenWriterVector);

    zserio::BitStreamReader reader(writtenWriterBuffer, writtenWriterBufferByteSize);
    const MetresConverterCaller readMetresConverterCaller(reader);
    ASSERT_EQ(metresConverterCaller, readMetresConverterCaller);
}

} // namespace structure_param
} // namespace functions
